File: src/lib/tiposkripto/src/verbs/tdt/BaseConfirm.ts
import type { TestTypeParams_any } from "../../CoreTypes.js";
import type { ITestArtifactory, ITestResourceConfiguration } from "../../types.js";
import { CommonUtils } from "../internal/CommonUtils.js";

/**
 * BaseConfirm for TDT (Table-Driven Testing) pattern.
 * Standalone class similar to BaseGiven but for table-driven testing.
 */
export class BaseConfirm<I extends TestTypeParams_any> {
  features: string[];
  testCases: any[][]; // Array of [Value, Should, Expected] tuples
  confirmCB: I["given"];
  initialValues: any;
  key: string = "";
  failed: boolean = false;
  artifacts: string[] = [];
  fails: number = 0;
  status: boolean | undefined;
  error: Error | null = null;
  store: I["istore"] = null as any;

  constructor(
    features: string[],
    testCases: any[][],
    confirmCB: I["given"],
    initialValues: any,
  ) {
    this.features = features;
    this.testCases = testCases || [];
    this.confirmCB = confirmCB;
    this.initialValues = initialValues;
  }

  addArtifact(path: string) {
    CommonUtils.addArtifact(this.artifacts, path);
  }

  setParent(parent: any) {
    (this as any)._parent = parent;
  }

  toObj() {
    const testCases = this.testCases || [];
    return CommonUtils.toObj(this, {
      testCases: testCases.map((testCase) => {
        if (Array.isArray(testCase) && testCase.length >= 2) {
          const [value, should] = testCase;
          
          // Get input data
          let inputData: any = null;
          try {
            if (typeof value === 'function') {
              inputData = value();
            } else if (value && typeof value.toObj === 'function') {
              const obj = value.toObj();
              inputData = obj.features || obj;
            } else {
              inputData = value;
            }
          } catch (e) {
            inputData = `Error: ${e.message}`;
          }
          
          // Get test description
          let testDescription: any = null;
          try {
            if (should) {
              if (typeof should === 'function') {
                // Try to get a better name from the function
                // For should functions that are created by beEqualTo(expected), etc.
                // They might have properties that indicate what they are
                if (should.name && should.name !== '') {
                  testDescription = should.name;
                } else {
                  // Try to extract from the function's string representation
                  const funcStr = should.toString();
                  if (funcStr.includes('beEqualTo')) {
                    testDescription = 'beEqualTo';
                  } else if (funcStr.includes('beGreaterThan')) {
                    testDescription = 'beGreaterThan';
                  } else if (funcStr.includes('whenMultipliedAreAtLeast')) {
                    testDescription = 'whenMultipliedAreAtLeast';
                  } else if (funcStr.includes('equal')) {
                    testDescription = 'equal';
                  } else {
                    testDescription = 'Test function';
                  }
                }
              } else if (should && typeof should.toObj === 'function') {
                const obj = should.toObj();
                testDescription = obj.name || 'Should';
              } else {
                testDescription = String(should);
              }
            }
          } catch (e) {
            testDescription = `Error: ${e.message}`;
          }
          
          return {
            input: inputData,
            test: testDescription,
          };
        }
        return testCase;
      }),
    });
  }

  async confirm(
    subject: I["isubject"],
    key: string,
    testResourceConfiguration: ITestResourceConfiguration,
    tester: (t: Awaited<I["then"]> | undefined) => boolean,
    artifactory?: any,
    suiteNdx?: number,
  ) {
    this.key = key;
    this.fails = 0;
    this.testResourceConfiguration = testResourceConfiguration;

    // Store suite index for use in artifactory creation
    (this as any)._suiteIndex = suiteNdx;

    // Create a proper artifactory if one isn't provided
    const actualArtifactory = artifactory;

    // For TDT tests, we don't need to setup a store via adapter.prepareEach
    // The confirmCB is the function to test (e.g., new Calculator().add)
    this.store = null;
    this.status = true;

    try {
      // Process each test case
      for (const [caseIndex, testCase] of this.testCases.entries()) {
        try {
          // Each test case is now [Value, Should] where Should is already called with expected value
          if (Array.isArray(testCase) && testCase.length >= 2) {
            const [value, should] = testCase;
            console.log('[BaseConfirm] value:', value);
            console.log('[BaseConfirm] should:', should);
            console.log('[BaseConfirm] value type:', typeof value);
            console.log('[BaseConfirm] should type:', typeof should);

            // Get the input from value
            let input: any;
            if (typeof value === 'function') {
              input = value();
              console.log('[BaseConfirm] input from function:', input);
            } else {
              // assume value is the input directly
              input = value;
              console.log('[BaseConfirm] input direct:', input);
            }
            
            // should is a function that expects the actual result
            if (typeof should === 'function') {
              // Log for debugging
              console.log('[BaseConfirm] input:', input);
              console.log('[BaseConfirm] confirmCB:', this.confirmCB);
              console.log('[BaseConfirm] should function:', should);
              // Compute actual result using confirmCB
              // For TDT, confirmCB might be:
              // 1. A function that returns the test function: () => (a, b) => a + b
              // 2. The test function itself: (a, b) => a + b
              let testFn;
              if (typeof this.confirmCB === 'function') {
                // Try to get the test function
                const potentialTestFn = this.confirmCB();
                if (typeof potentialTestFn === 'function') {
                  testFn = potentialTestFn;
                } else {
                  // this.confirmCB is the test function itself
                  testFn = this.confirmCB;
                }
              } else {
                testFn = this.confirmCB;
              }
            
              // Now call the test function with input
              const actualResult = Array.isArray(input) 
                ? testFn(...input) 
                : testFn(input);
              console.log('[BaseConfirm] actualResult:', actualResult);
              // Call should with the actual result
              const passed = should(actualResult);
              tester(passed);
            } else if (should && typeof should.processRow === 'function') {
              // should is a BaseShould instance
              // Compute actual result using confirmCB
              const actualResult = Array.isArray(input) 
                ? this.confirmCB(...input) 
                : this.confirmCB(input);
              console.log('[BaseConfirm] actualResult:', actualResult);
              const passed = await should.processRow(
                actualResult,
                testResourceConfiguration,
                artifactory,
              );
              tester(passed);
            } else {
              // No validation performed
              tester(true);
            }
          }
        } catch (e: any) {
          CommonUtils.handleTestError(e, this);
        }
      }
    } catch (e: any) {
      CommonUtils.handleTestError(e, this);
    }

    return this.store;
  }

  // Alias for run to match BaseSuite expectations
  async run(
    subject: I["isubject"],
    testResourceConfiguration: ITestResourceConfiguration,
    artifactory?: ITestArtifactory,
  ) {
    return this.confirm(
      subject,
      this.key || "confirm",
      testResourceConfiguration,
      (t) => !!t,
      artifactory,
    );
  }
}

export type IConfirms<I extends TestTypeParams_any> = Record<string, BaseConfirm<I>>;

File: src/lib/rubeno/lib/base_confirm.rb
module Rubeno
  # BaseConfirm for TDT (Table-Driven Testing) pattern.
  # Standalone class similar to BaseGiven but for table-driven testing.
  class BaseConfirm
    attr_accessor :features, :test_cases, :confirm_cb, :initial_values, :key, :failed, :artifacts, :fails, :status, :error, :store,
                  :test_resource_configuration
    
    def initialize(features, test_cases, confirm_cb, initial_values)
      @features = features
      @test_cases = test_cases || []
      @confirm_cb = confirm_cb
      @initial_values = initial_values
      @artifacts = []
      @fails = 0
      @failed = false
      @error = nil
      @store = nil
      @key = ""
      @status = nil
      @test_resource_configuration = nil
    end
    
    def add_artifact(path)
      normalized_path = path.gsub('\\', '/')
      @artifacts << normalized_path
    end
    
    def set_parent(parent)
      @_parent = parent
    end
    
    def to_obj
      {
        key: @key,
        features: @features,
        test_cases: @test_cases,
        failed: @failed,
        fails: @fails,
        error: @error ? "#{@error.class}: #{@error.message}" : nil,
        artifacts: @artifacts,
        status: @status
      }
    end
    
    def confirm(subject, key, test_resource_configuration, tester, artifactory = nil, suite_ndx = nil)
      @key = key
      @fails = 0
      @test_resource_configuration = test_resource_configuration
      
      # Store suite index for use in artifactory creation
      @_suite_index = suite_ndx
      
      # For TDT tests, the confirm_cb is the function to test
      @store = nil
      @status = true
      
      begin
        # Process each test case
        @test_cases.each_with_index do |test_case, case_index|
          begin
            if test_case.is_a?(Array) && test_case.length >= 2
              value, should = test_case
              
              # Get the input from value
              input = value.respond_to?(:call) ? value.call : value
              
              # Compute actual result using confirm_cb
              test_fn = @confirm_cb
              if test_fn.respond_to?(:call)
                actual_result = if input.is_a?(Array)
                  test_fn.call(*input)
                else
                  test_fn.call(input)
                end
                
                # Call should with the actual result
                if should.respond_to?(:call)
                  passed = should.call(actual_result)
                  tester.call(passed)
                elsif should.respond_to?(:process_row)
                  passed = should.process_row(actual_result, test_resource_configuration, artifactory)
                  tester.call(passed)
                else
                  tester.call(true)
                end
              end
            end
          rescue => e
            @failed = true
            @fails += 1
            @error = e
          end
        end
      rescue => e
        @failed = true
        @fails += 1
        @error = e
      end
      
      @store
    end
    
    # Alias for run to match BaseSuite expectations
    def run(subject, test_resource_configuration, artifactory = nil)
      confirm(subject, @key || "confirm", test_resource_configuration, ->(t) { !!t }, artifactory)
    end
  end
end
